src/session.c

```diff
@@ -7,33 +7,168 @@
+        pool->limit_sessions = pool->max_sessions;
+    }
+
+    *_pool = pool;
+    return CKR_OK;
+}
+
+CK_RV p11prov_session_pool_free(P11PROV_SESSION_POOL *pool)
+{
+    P11PROV_SESSION *session;
+
+    if (!pool) {
+        return CKR_OK;
+    }
+
+    /* deref login_session first */
+    p11prov_session_free(pool->login_session);
+
+    /* LOCKED SECTION ------------- */
+    pthread_mutex_lock(&pool->lock);
+    for (int i = 0; i < pool->num_sessions; i++) {
+        session = pool->sessions[i];
+        if (session->refcnt > 1) {
+            P11PROV_raise(pool->provctx, CKR_GENERAL_ERROR,
+                          "Can't free multiply-referenced session (%u)", i);
+        }
+        if (session->session != CK_INVALID_HANDLE) {
+            CK_FUNCTION_LIST *f;
+            int ret;
+
+            f = p11prov_ctx_fns(pool->provctx);
+            if (f != NULL) {
+                ret = f->C_CloseSession(session->session);
+                if (ret != CKR_OK) {
+                    P11PROV_raise(pool->provctx, ret,
+                                  "Failed to close session %lu",
+                                  session->session);
+                }
+            }
+            (void)__atomic_fetch_sub(&pool->cur_sessions, 1, __ATOMIC_SEQ_CST);
+        }
+        OPENSSL_clear_free(session, sizeof(P11PROV_SESSION));
+        pool->sessions[i] = NULL;
+    }
+    OPENSSL_free(pool->sessions);
+    pool->sessions = NULL;
+    pool->num_sessions = 0;
+    pthread_mutex_unlock(&pool->lock);
+    /* ------------- LOCKED SECTION */
+
+    pthread_mutex_destroy(&pool->lock);
+    OPENSSL_clear_free(pool, sizeof(P11PROV_SESSION_POOL));
+
+    return CKR_OK;
+}
+
+#define SESS_ALLOC_SIZE 8
+static P11PROV_SESSION *p11prov_session_new(P11PROV_CTX *ctx,
+                                            struct p11prov_slot *slot)
 {
-    P11PROV_SESSION *sess;
+    P11PROV_SESSION_POOL *p = slot->pool;
+    P11PROV_SESSION *session;
+    CK_ULONG cur_sessions;
+
+    /* this is just to quit early and avoid work.
+     * Passing this check does not guarantee another thread is not
+     * racing to open the last avaliable session */
+    cur_sessions = __atomic_load_n(&p->cur_sessions, __ATOMIC_SEQ_CST);
+    if (cur_sessions >= p->max_sessions) {
+        P11PROV_raise(ctx, CKR_SESSION_COUNT, "Max sessions limit");
+        return NULL;
+    }
 
-    sess = OPENSSL_zalloc(sizeof(P11PROV_SESSION));
-    if (sess == NULL) {
+    session = OPENSSL_zalloc(sizeof(P11PROV_SESSION));
+    if (session == NULL) {
         P11PROV_raise(ctx, CKR_HOST_MEMORY, "Failed to allocate session");
         return NULL;
     }
 
-    sess->provctx = ctx;
-    sess->slotid = slotid;
-    sess->session = CK_INVALID_HANDLE;
+    session->provctx = ctx;
+    session->slotid = slot->id;
+    session->session = CK_INVALID_HANDLE;
+    session->pool = p;
+    session->refcnt = 1;
+
+    /* LOCKED SECTION ------------- */
+    pthread_mutex_lock(&p->lock);
+
```

Comment:
Maybe check here for `p->num_sessions >= p->max_sessions`. Otherwise due to races, it's possible allocate more `P11PROV_SESSION` objects than allowed.

Reply:
Not sure it is really worth it, at most we waste 64/256 bytes of data anyway.

Reply:
Yeah this ties in to the other comment I just left.
Should we have a limit for P11PROV_SESSIONS? Or should we allow a runaway thread to just allocate as many as it wants?
Perhaps we should allow just a handful more than max_sessions ?